### ISSUE TITLE
Log the log events in twisted.internet.protocol.Factory using the new logging system

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7897]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)| @eeshangarg reported|
|-|-|
|Trac ID|trac#7897|
|Type|enhancement|
|Created|2015-05-13 04:42:14Z|
|Branch|https://github.com/twisted/twisted/tree/factory-log-7897|

Attachments:

* [feature_7897.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b2d/b2d0cc54e5a2edd06277d5888bd42a879627b41e/e6d3802cfdf8c1801b4883c1ec7af8bd1a091a59.patch) (3690 bytes) - added by eeshangarg on 2015-05-13 05:13:15Z - 
* [feature_7897_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b2d/b2d0cc54e5a2edd06277d5888bd42a879627b41e/c6d18856dff0065d45e3e2b8c698dd37753c736f.patch) (2982 bytes) - added by eeshangarg on 2015-05-22 07:32:53Z - 
* [feature_7897_3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b2d/b2d0cc54e5a2edd06277d5888bd42a879627b41e/00671b9d578b1c01f1ae7a11b176fdd415a82ddb.patch) (525 bytes) - added by eeshangarg on 2015-05-22 14:17:34Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7897 7897
type__enhancement enhancement
reporter__eeshangarg eeshangarg
priority__normal normal
milestone__ 
branch__branches_factory_log_7897 branches/factory-log-7897
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1431492134695679 1431492134695679
changetime__1433653409298010 1433653409298010
version__None None
owner__glyph glyph
cc__eeshangarg
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)|@eeshangarg set owner to @eeshangarg|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_2"></a>|@eeshangarg removed owner|
|-|-|

Hi!

I just submitted a patch for this (feature_7897.patch). However, there are a couple of issues that my patch does not address and I also ran into a few problems (well, not problems per se, just things that, in my opinion, warranted a longer discussion); as a consequence of which I have some concerns that I will try my best to summarize in this ticket comment.

In my patch, I added an attribute called "log" to the Factory class which instantiates a Logger without passing an observer or anything else. If you do not include this "log" attribute, anything like:

```
self.log.info("Starting factory {factory!r}", factory=self)
```

will raise AttributeError. It is apparent that this approach has a few caveats. First of all, if a person decides to create a Logger globally for a module like so <http://twisted.readthedocs.org/en/latest/core/howto/logger.html#usage-for-emitting-applications>, then referring to this global Logger as self.log will be logging events to the Logger that was instantiated in the Factory class's log attribute (which may or may not exist!) and not to the Logger created globally for an entire module (which also may or may not exist!).

The second thing is, even if we do create a Logger globally for the whole twisted.internet.protocol module (so that references to anything like "log.info" do not raise an AttributeError exception), then we obviously can't refer to it as self.log from within any class. Another side effect of this approach is that if a person wants to create a Logger and associate it with a particular class, then logging statements within that particular class, for instance:

```
log.info("Starting factory {factory!r}", factory=self)
```

will be logging events to the globally-created Logger and not to the Logger in the class's log attribute (because we need to access it using "self.log" and not just "log.info" or something). I can be wrong, but to me it seems as if trying to circumvent any of the above-mentioned limitations involves writing code the sole purpose of which would be to basically figure out which Logger to log it's events to.

Off the top of my head, within a class, I think we could do something like:

```
log = None

if self.log is not None:
    # if the log attribute in the class was assigned a Logger 
    self.log.info("Starting factory {factory!r}", factory=self)
else:
    try:
        # else try to log to a Logger created globally for a module
        log.info("Starting factory {factory!r}", factory=self)
    except AttributeError:
        # fall back to the old logging system (at least up until the point when
        # we can move to the new system and start using it exclusively) 
        log.msg("Starting factory %r" % self")
```

Admittedly, this code snippet is just a hasty workaround, and I am aware that this is probably not the recommended way to solve the problem at hand and something tells me that the above snippet (or anything similar) is a highly unrealistic way of going about this, taking into account that the move to the new logging system may span hundreds of tickets/patches modifying a lot of logging statements; hence, my decision to seek advice here.

I would really appreciate it if someone could share their insights and give me some guidance as to how to move further with this. :-)

Thanks!
Eeshan

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

Thanks for working on this.

To put this into context: 

Logging consists of two main endpoints: applications that emit events, and observers that receive and handle those events.

As twisted core developers I think that we will only implement the `emit events` parts but we should consider a few observers use cases.

-------- 

I will try to describe my use case and maybe this would help with the progress of this ticket.

I have a number of independent / distributed python twisted processes.
The logs are also aggregated into a single place... but also stored on local servers.
A single python/twisted process can run multiple factories.

Each emitted log has a standard set of key/values: event UID / process UUID / factory UUID / connection UUID / remote peer address / authentication account / severity. This is similar to the idea of 'Event keys added by the system'

Logs are emitted using a hierarchical call to fill the structured log data. Protocol class will add the connection UUID and remote peer / the factory class will add factory UUID and the process class will add the process UUID. The process class will also handle the actual dispatching of the event using a set of observers.

To help the support team, each emitted log has an unique ID. Each unique ID has an associated format text, version in which it was added / removed and some notes/comments about the conditions in which the event is emitted.

We create a separate HTML page which documents all possible events which can be raised by our application. In this way 3rd parties can just read the documentation and will know what events they should observe.
Also the support team don't have to read the source code to find details about a log message.

For my use case the code look like

```
class SomeFactory():

    def init(self, parent, uuid):
        self._parent = parent
        self._uuid = uuid

    def somethingHappend(self):
        self.emit(id=1234, text='"Starting factory {factory!r}"', factory=self)

    def emit(self, id, **kwargs):
        self._parent.emit(id, factory_uuid=self._uuid, **kwargs)

```

-------------

In this case I think that this is fine

```
self.log.info("Starting factory {factory!r}", factory=self)

or

self.log("Starting factory {factory!r}", factory=self, level='info')
```

each factory as init/build time can set up its own `self.log` method

--------

With the existence of the globalLogPublisher (https://twistedmatrix.com/documents/15.2.0/api/twisted.logger.html#globalLogPublisher) I think that this is solved and the current code is good.


-----------

My comments for this review are:

the way the log class instance is defined and the way log is called are good.

For the test code, I think that  we should start creating an assertion helper for log events as otherwise each time we need to assert that an event was emitted we will have to duplicate all those many lines to check for formatting, log level, dict data.



Instead of

```
self.assertIs(events[0]['factory'], f)
self.assertEqual(events[0]['log_level'], LogLevel.info)
self.assertEqual(events[0]['log_format'],
    'Stopping factory {factory!r}')
self.assertEqual(
    formatEvent(events[0]),
    u'Stopping factory {factory!r}'.format(factory=f)
    )
```

we can have:

```

self.assertEvent(
    events[0],
    log_level=LogLevel.info,
    log_format='Stopping factory {factory!r}',
    factory=f)

def assertEvent(event, log_level, log_format, **kwargs):
    self.assertEqual(event['log_level'], log_level)
    self.assertEqual(event['log_format'], log_format)
    self.assertEqual(formatEvent(event), log_format.format(**kwargs))
    for key, value in kwargs.items():
        self.assertIs(event['key'], value)

```


In the future we might also want to have this assertion to make sure that the
code will emit the exact/strict number of events

```
self.assertEvents([
    {
    log_level=LogLevel.info,
    log_format='Starting factory {factory!r}',
    other_key=value,
    },
    {
    log_level=LogLevel.info,
    log_format='Stopping factory {factory!r}',
    factory=f,
    },
    ], events)
```

as with the current code we might want to update the test code to assert that only a single event is emitted.
In case of a bug and the event is emitted twice the current test will still pass.


-------

Many thanks for working on this!

Please note that I am also new to the new logging system and I don't know to many things about its design.

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_4"></a>|@moshez commented|
|-|-|

Adi -- you didn't remove the review keyword. Is this on purpose?

Also:
I'm not really sure why assertEvent is useful
1. The formatting stuff is really testing formatEvent over and over again. The test for formatEvent should cover formatEvent :)
2. If you take the formatting out of the equation, you're left with something not much more than assertEquals(events[0], dict(loglevel=LogLevel.info, log_format='Starting ...', factory=f)) or assertEquals(events, [dict(loglevel=LogLevel.info, log_format='Starting ...', factory=f), dict(...log_format='Stopping...)]) which are no longer than using assertEvent

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

(In [44904]) Branching to factory-log-7897.

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_6"></a>|@eeshangarg commented|
|-|-|

Hi!

I just added a new patch called feature_7897.patch created on top of the factory-log-7897 branch. This patch removes the assertions that use formatEvent in the tests, after reading moshez's comments and after a discussion on IRC, I don't think the formatEvent assertions will fail if everything that precedes them passes, so they don't really need to be there. Plus, this patch also corrects a minor mistake in the topfile pointed out by hawkowl. And this makes the "log" attribute in twisted.internet.protocol.Factory private ("_log") so that calls in the form of "self.log" do not clash with this <https://github.com/twisted/twisted/blob/ab110cc815c31eed17bb5ef06f1037a8ede0b395/twisted/web/http.py#L2072> and this <https://github.com/twisted/twisted/blob/38421d6fcffa1ddb590e51df0e1c6cba6f29d052/twisted/protocols/socks.py#L183>, in the event of such a clash, dozens of tests would fail, throwing the following exception, (see <http://codepad.org/6UlkCuHj>):

```
exceptions.AttributeError: 'function' object has no attribute 'info'

twisted.web.test.test_xmlrpc.XMLRPCPublicLookupProcedureTests.test_lookupProcedure
```

Let me know if this needs further work! If so, I would love to work on it. :-)

Thanks to adiroiban, hawkowl, moshez and tomprince for their help and guidance!

Regards,
Eeshan

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

I did not removed the review on purpose, so that at least another developer will check the changes.

I don't know all the details of the new log system but I assume it does lazy text resolution. This mean that unless we don't have an observer the text is not resolved.

I would rather remove this assertion and replace it with formatEvent, as the assertion will most probably be written with a copy paste ... either copy from test to code or copy for code to test.

```
self.assertEqual(events[0]['log_format'],
    'Stopping factory {factory!r}')
```

----------

Another option is to use a real observer to help with testing, instead of the simple `observer=events.append`

The test dedicated observer can do its own format check and provide helper methods to assert the events.

I am +1 for a better test observer, but I think that this can be done in a separate ticket.

So yes, we can leave this code as it is and improve our test helpers in a separate ticket.


---------

I am also not removing the 'review' this time so that people could still see it and add their comments

Will apply the new patch and sent it to builders


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

there is one pyflakes error left: 
https://buildbot.twistedmatrix.com/builders/pyflakes/builds/2670/steps/pyflakes/logs/new%20pyflakes%20errors

twisted code should be pretty clean of pyflakes error so you can run pyflakes on your local computer and check for pyflakes errors

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_9"></a>|@eeshangarg commented|
|-|-|

Hi adiroiban! 

I have submitted a patch that fixes the pyflakes error. I ran pyflakes before submitting this patch and I also ran the tests again. From now on I'll make sure that I run pyflakes before submitting any patch. :-)

Thanks a lot for your help!

Regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

I'm glad that we ran into the `self.log`-being-a-public-attribute thing early on, with the conflicts that you mentioned.  I think it should be a recommended practice going forward, since we rarely, if ever, want an object's logging interface to be part of its public API.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set owner to @glyph|
|-|-|

# The Road So Far

To summarize the sequence of events thus far so others don't have to read through the full comment history:

    1. Eeshan wrote a patch that did testing via comparing the output of `.format` on some test data with `formatEvent` on the logged data, and asked some questions about it.  (Since these questions are largely resolved now I won't repeat them here.)
    2. Adi answered the questions and made some suggestions, including a suggestion to add an `assertEvent`.
    3. Moshe did a code review and suggested that calling `formatEvent` was pointless, since we're not testing `formatEvent` here.
    4. Eeshan wrote another patch that compared the logged data instead to static values.

# What Next

Here are some feedback points which could be addressed in subsequent changes.

    1. The current implementation is not testing that the emitted data structures are the proper shape to integrate with anything within Twisted.  I preferred comparing to `formatEvent` to what the newer patch does, because at least `formatEvent` is a source of some authority, outside the test, which makes sure that the things logged look like log events.  However, _what_ was being compared to `formatEvent` was some code that copy/pasted format strings and compared the result of calling `.format()` on them, so that's not ideal either.
    1. Another issue with the current implementation is that it is not verifying the integration between the system under test and the way it satisfies its own dependencies: the `_log` attribute supplied by the test subclass _might not be supplied by the system under test_; an implementation which failed to initialize `_log` at all would still pass these tests.  Adi's `assertEvent` suggestion could automate this, providing an appropriate level of integration with the existing log system and carefully managing any global state mutations.
    1. Nothing ever asserts about the ''length'' about the returned list of events; `doStart`, as far as we know, should log one and only one thing, and if it logs more things that would be a surprise.  When testing stuff like this, I frequently do something like "`[event] = events`" to make sure I'm not cluttering up my tests with lots of overly-verbose asserts, but that I'll still notice if my expectation about the number of calls is wrong.

I think we should discuss more about how to ensure that state-modifications are as local as possible to avoid contaminating shared state either in tests or the implementation, and yet still using realistic APIs everywhere possible so if things change or are deprecated callers will notice.

One idea that comes to mind is to leverage the fact that `_log` is a descriptor which returns a new object.  So you could, for example, do something like this:

```
#!py
# implementation.py
class ThingThatLogs(object):
    _log = Logger()
# test_implementation.py
class ThingThatLogsTests(TestCase):
    def test_thingThatLogs(self):
        unitUnderTest = ThingThatLogs()
        aLogger = unitUnderTest._log
        loggedStuff = []
        aLogger.observer = loggedStuff.append
        unitUnderTest._log = aLogger
```

In other words, peel the descriptor-ized instance-level `Logger` off the instance, then put it back so that a new one won't be created next time, with a modified `observer`. This way the test code isn't satisfying the direct dependency (on a `Logger` that should have been initialized in the code itself) but instead is satisfying the dependency one level down.

Then there's still the question about how to ensure the object emitted is a proper log event.  I would prefer to use something explicit like Adi's proposed `assertEvent` rather than just calling `formatEvent` which just implicitly inspects only some of the expected event API surface area.

# Summing Up

I think that the branch as it stands is a good initial improvement, and the test coverage is _sufficient_, so I will land it.  However, the test coverage is not _optimal_, so I would appreciate it if the issues above were addressed in follow-up tickets.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set status to `closed`|
|-|-|

(In [45008]) Merge factory-log-7897: use logger from Factory

Author: eeshangarg

Reviewer: glyph

Fixes: [#7897](https://github.com/twisted/twisted/issues/7897)

twisted.internet.protocol.Factory now uses the new logging
system (twisted.logger) for all its logging statements.
